File: 1218.longest-arithmetic-subsequence-of-given-difference.cpp
/*
 * @lc app=leetcode id=1218 lang=cpp
 *
 * [1218] Longest Arithmetic Subsequence of Given Difference
 */
#include <vector>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
    int base = 20000;

public:
    int longestSubsequence(vector<int> &arr, int difference)
    {
        vector<int> prev_diff_max_count(40001);
        int res = 1;
        for (auto n : arr)
        {
            prev_diff_max_count[n + base] = prev_diff_max_count[n - difference + base] + 1;
            res = max(res, prev_diff_max_count[n + base]);
        }
        return res;
    }
};
// @lc code=end

File: py/1218.longest-arithmetic-subsequence-of-given-difference.py
#
# @lc app=leetcode id=1218 lang=python3
#
# [1218] Longest Arithmetic Subsequence of Given Difference
#
from collections import Counter
# @lc code=start


class Solution:
    def longestSubsequence(self, arr: List[int], difference: int) -> int:
        prev_diff_max_count = Counter()
        res = 1
        for n in arr:
            prev_diff_max_count[n] = prev_diff_max_count[n-difference]+1
            res = max(res, prev_diff_max_count[n])
        return res
# @lc code=end
